cpp/tensorrt_llm/thop/attentionOp.cpp

```diff
@@ -237,22 +237,66 @@ class Runner : public RunnerBase
                     ? host_kv_cache_block_offsets.value().index({pool_index, seq_offset}).data_ptr()
                     : nullptr);
 
-        auto const cache_elem_size = (op.mKVCacheQuantMode.hasKvCacheQuant() ? 1 : sizeof(T));
+        // The cache element size in bits.
+        int cache_elem_bits;
+        if (op.mKVCacheQuantMode.hasInt8KvCache() || op.mKVCacheQuantMode.hasFp8KvCache())
+        {
+            cache_elem_bits = 8;
+        }
+        else if (op.mKVCacheQuantMode.hasFp4KvCache())
+        {
+            cache_elem_bits = 4;
+        }
+        else
+        {
+            cache_elem_bits = sizeof(T) * 8;
+        }
         auto const block_size = op.mTokensPerBlock * op.mNumKVHeads * op.mHeadSize;
-        auto const bytes_per_block = block_size * cache_elem_size;
+        auto const bytes_per_block = block_size * cache_elem_bits / 8 /*bits*/;
         int32_t const kv_factor = op.isMLAEnabled() ? 1 : 2;
         auto const intra_pool_offset = layer_idx_in_cache_pool * kv_factor * bytes_per_block;
 
-        void* host_primary_pool_pointer = op.useKVCache() && host_kv_cache_pool_pointers.has_value()
-            ? reinterpret_cast<void*>(
+        // Prepare block pool pointers for NVFP4 KV cache.
+        void* host_primary_pool_pointer{nullptr};
+        void* host_secondary_pool_pointer{nullptr};
+        void* host_primary_block_scale_pool_pointer{nullptr};
+        void* host_secondary_block_scale_pool_pointer{nullptr};
+
+        // Whether NVFP4 KV cache is used.
+        bool const use_kv_cache = op.useKVCache() && host_kv_cache_pool_pointers.has_value();
+        bool const use_nvfp4_kv_cache = use_kv_cache && op.mKVCacheQuantMode.hasFp4KvCache();
+        if (use_nvfp4_kv_cache)
+        {
+            // For NVFP4 KV cache, extra block scales are stored in separate pools.
+            // The layout of host_kv_cache_pool_pointers is [num_pools, 2 (primary and secondary), 2 (data and scale)].
+            TORCH_CHECK(host_kv_cache_pool_pointers.value().dim() == 3);
+            host_primary_pool_pointer = reinterpret_cast<void*>(
+                reinterpret_cast<char*>(host_kv_cache_pool_pointers.value().index({pool_index, 0, 0}).item<int64_t>())
+                + intra_pool_offset);
+            host_secondary_pool_pointer = reinterpret_cast<void*>(
+                reinterpret_cast<char*>(host_kv_cache_pool_pointers.value().index({pool_index, 1, 0}).item<int64_t>())
+                + intra_pool_offset);
+            // Calculate the intra-pool offset for scaling factors.
+            auto constexpr vector_size = 16;
```

Comment:
we use a fixed vector size, right ? probably leave a comment or something.

Reply:
Done.